README
Crouching Process Hidden File (cphf)
https://github.com/x5Fhayling/cphf
Spilling the details on unlinked files and the processes accessing them

When a file is deleted it usually isn't erased. Rather, it's more like its entry
in the directory is deleted (unlinked). What can happen though is even if that
entry is deleted the file could still be in use by a process or program and the
space the file is using won't be freed for reuse until that program is done with
it. While this is a legit technique for making sure temp files get deleted once a
program ends, it is also used by malware and malicious hackers.

cphf goes through and finds those unlinked but still in use files on a linux
system and spills the deets on the files and the processes accessing them. :D


Usage: cphf [--file no|acdgimMnpsu] [--proc no|cdDegGiMnPqrsSTuUv!]
NOTE: running cphf under different privileges will likely return different results

--file	f	Suppresses files from being listed or specifies which
	   	attributes should be listed.  Options are:
		no 	Supress all file information
		a	Last time file was accessed
		c	Last time status was changed
		d	Device file resides on
		g	Group ID
		i	INode
		m	Last modification time
		M	File Mode
		n	File Name
		p	Permissions
		s	Size (bytes)
		u	User ID

--proc 	p	Supresses processes from being listed or specifies which
		attributes should be listed.  Options are:
                no      Supress all process information
		c	Command with options
		d	Device ID
		D	Process Data Region Memory Address range
		e	Process start time
		g	Group ID
		G	Group Name
		i	Instruction Pointer
		M	Permissions
		n	Process Group ID
		P	Parent Process ID
		q	CPU Time
		r	Resident Set (pages)
		s	State of the process
		S	Process Stack Memory Address range
		T	Process Text Region Memory Address range
		u	User ID
		U	Login name
		v	Virtual Memory (bytes)


Sample Output:

Command                  /tmp/tcpdump -i eth0 -w /tmp/capture 
Process ID (PID)         2320
Device                   8,1	?
Process Group ID (PGID)  2319
Parent PID               2319
User Name                root
User ID (UID)            0
User Group Name          root
User Group ID (UGID)     0
Current State            S (Sleeping in an interruptible wait)
Permissions              -rwxr-xr-x
Start Time               Thu Jun 25, 2015 12:35:47 PDT
CPU Time                 0:00:00
Virtual Memory (bytes)   19615744
Resident Set (pages)     1411
Text Region              0x7f90fc075000  -  0x7f90fc14e9f4
Data Region              0x7f90fc36e480  -  0x7f90fdb49000
Stack                    0x7ffe0bcd7aa0  -  0x7ffe0bcd6698
Instruction Pointer      0x7f90fb55f110

	Filename                   /tmp/capture (deleted)
	Device                     8,1	?
	Inode                      777
	User ID (UID)              0
	User Group ID (UGID)       0
	File Mode                  Regular File
	Permissions                -rw-r--r--
	Last Accessed              Thu Jun 25, 2015 12:35:48 PDT
	Last Modified              Thu Jun 25, 2015 12:37:09 PDT
	Last Status Change         Thu Jun 25, 2015 12:37:09 PDT
	File Size (bytes)          69632

	Filename                   /tmp/tcpdump (deleted)
	Device                     8,1	?
	Inode                      347
	User ID (UID)              0
	User Group ID (UGID)       0
	File Mode                  Regular File
	Permissions                -rwxr-xr-x
	Last Accessed              Thu Jun 25, 2015 12:35:48 PDT
	Last Modified              Thu Jun 25, 2015 12:35:39 PDT
	Last Status Change         Thu Jun 25, 2015 12:36:50 PDT
	File Size (bytes)          1023160


NEWS
25 June 2015 	Welcome to the first alpha release of cphf! Right now output 
		options are limited to "report mode" however more digestible 
		output is on the docket.  Additionally right now cphf calls 
		out to lsof to find unlinked files. This will change in 
		future releases.  Up next is an overhaul of how attributes
		are being stored and handled. Also I want to go through all
		of the primitives and make sure no overflows are possible.
		If you find any bugs or have any suggestions please drop a line
		at github repository: https://github.com/x5Fhayling/cphf  


src/CPHFCommon.h
///////////////////////////////////////////////////////////////////////////////
//    Crouching Process Hidden File (cphf)
//    Spilling the details on unlinked files and the processes accessing them
//    Copyright (C) 2015 Erin Adamek
//    Please report bugs to: https://github.com/x5Fhayling/cphf/issues
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
///////////////////////////////////////////////////////////////////////////////


#ifndef CPHFCOMMON_H
#define CPHFCOMMON_H

#include <sstream>
#include <map>
#include <unordered_map>
#include <vector>
#include <utility>  //std::pair
#include <stdio.h>  //printf
#include <memory>   //shared_ptr used for std::vector<CPHFFile*>
                    //      to prevent memory leaks

#include <linux/version.h>
#include <unistd.h>


#define CPHF_SUCCESS 1
#define CPHF_ERROR -1

//#define BUFFERSIZE  1024


       
#define CPHF_FORMAT_DATE_TIME "%a %b %d, %Y %T %Z"
#define CPHF_FORMAT_TIME  "%d:%02d:%02d"
//attriute pairs take the form: CPHF_ID, value
using cphf_attribute = std::pair<int, std::string>;



//List of attribute IDs and Labels



// lsof defined

enum cphf_ids{  CPHF_ID_PID         =   5,
                CPHF_ID_FD          =   105,
                CPHF_ID_LOCK		=   130,
                CPHF_ID_CMD         =   0,
                CPHF_ID_FILENAME    =   2,
                CPHF_ID_DEVICE      =   6,
                CPHF_ID_INODE		=   8,
                CPHF_ID_PROCGID     =   10,
                CPHF_ID_PPID        =   15,
                CPHF_ID_LOGINNAME   =   20,
                CPHF_ID_UID         =   25,
                CPHF_ID_UGID_NAME   =   30,
                CPHF_ID_UGID        =   35,
                CPHF_ID_FILEMODE    =   39,
                CPHF_ID_PSTATE      =   40,
                CPHF_ID_FLAGS       =   41,
                CPHF_ID_PERMS       =   42,
                CPHF_ID_PSTART_TIME =   45,
                CPHF_ID_ACCESS_T    =   46,
                CPHF_ID_MOD_T       =   47,
                CPHF_ID_STAT_CHANGE_T=  48,
                CPHF_ID_PCPU_TIME   =   50,
                CPHF_ID_RAW_DEVICE  =   54,
                CPHF_ID_FILESIZE    =   57,
                CPHF_ID_PVMS        =   60,
                CPHF_ID_PRSS        =   65,
                CPHF_ID_TEXT_REGION =   70,
                CPHF_ID_DATA_REGION =   75,
                CPHF_ID_STACK_REGION=   80,
                CPHF_ID_ESP         =   85,
                CPHF_ID_EPI         =   90};


const std::string	CPHF_LABEL_PID              = "Process ID (PID)";
const std::string	CPHF_S_LABEL_PID            =   "PID";

const std::string	CPHF_LABEL_FD               =   "File Descriptor";
const std::string	CPHF_S_LABEL_FD             =   "FD";


const std::string   CPHF_LABEL_LOCK             =   "Lock (r|w|u)";
const std::string	CPHF_S_LABEL_LOCK           =   "FLOCK";

//  /roc defined

const std::string	CPHF_LABEL_CMD              =   "Command";
const std::string	CPHF_S_LABEL_CMD            =   "CMD";


const std::string	CPHF_LABEL_FILENAME         =   "Filename";  //fid
const std::string	CPHF_S_LABEL_FILENAME       =   "FNAME";

const std::string   CPHF_LABEL_DEVICE           =   "Device";
const std::string	CPHF_S_LABEL_DEVICE         =   "DEVID";



const std::string	CPHF_LABEL_INODE            =   "Inode";     //fid
const std::string	CPHF_S_LABEL_INODE          =   "INODE";

const std::string	CPHF_LABEL_PROCGID          =   "Process Group ID (PGID)";
const std::string	CPHF_S_LABEL_PGID           =   "PGID";

const std::string	CPHF_LABEL_PPID             =   "Parent PID";
const std::string	CPHF_S_LABEL_PPID           =   "PPID";

const std::string   CPHF_LABEL_LOGINNAME        =   "User Name";
const std::string	CPHF_S_LABEL_LOGINNAME      =   "UNAME";

const std::string   CPHF_LABEL_UID              =   "User ID (UID)";
const std::string	CPHF_S_LABEL_UID            =   "UID";

const std::string   CPHF_LABEL_UGID_NAME        =   "User Group Name";
const std::string	CPHF_S_LABEL_UGID_NAME      =   "UGN";

const std::string   CPHF_LABEL_UGID             =   "User Group ID (UGID)";
const std::string	CPHF_S_LABEL_UGID           =   "UGID";

const std::string	CPHF_LABEL_FILEMODE         =   "File Mode"; //fid
const std::string	CPHF_S_LABEL_FILEMODE       =   "MODE";

const std::string	CPHF_LABEL_PSTATE           =   "Current State";
const std::string	CPHF_S_LABEL_PSTATE         =   "STATE";

const std::string	CPHF_LABEL_FLAGS            =   "File Flags";
const std::string	CPHF_S_LABEL_FLAG           =   "FFLAG";

const std::string   CPHF_LABEL_PERMS            =   "Permissions";
const std::string	CPHF_S_LABEL_PERMS          =   "PERM";

const std::string   CPHF_LABEL_PSTART_TIME      =   "Start Time";
const std::string	CPHF_S_LABEL_PSTART_TIME    =   "STRTT";

const std::string   CPHF_LABEL_ACCESS_T         =   "Last Accessed";//fid
const std::string	CPHF_S_LABEL_ACCESS_T       =   "ACCT";

const std::string   CPHF_LABEL_MOD_T            =   "Last Modified"; //fid
const std::string	CPHF_S_LABEL_MOD_T          =   "MODT";

const std::string   CPHF_LABEL_STAT_CHANGE_T    =   "Last Status Change";//fid
const std::string	CPHF_S_LABEL_STAT_CHAGE_T   =   "STATT";


const std::string   CPHF_LABEL_PCPU_TIME        =   "CPU Time";
const std::string	CPHF_S_LABEL_PCUP_TIME      =   "CPUT";

const std::string   CPHF_LABEL_RAW_DEVICE       =   "Raw Device";
const std::string	CPHF_S_LABEL_RAW_DEVICE     =   "RDEV";

const std::string   CPHF_LABEL_FILESIZE         =   "File Size (bytes)"; //fid
const std::string	CPHF_S_LABEL_FILESIZE       =   "FSIZE";

const std::string   CPHF_LABEL_PVMS             =   "Virtual Memory (bytes)";
const std::string   CPHF_S_LABEL_PVMS           =   "VMS";

const std::string   CPHF_LABEL_PRSS             =   "Resident Set (pages)";
const std::string	CPHF_S_LABEL_PRSS           =   "RSS";

const std::string	CPHF_LABEL_TEXT_REGION      =   "Text Region";
const std::string	CPHF_S_LABEL_TEXT_REGION    =   "TEXTR";

const std::string   CPHF_LABEL_DATA_REGION      =   "Data Region";
const std::string   CPHF_S_LABEL_DATA_REGION    =   "DATAR";

const std::string	CPHF_LABEL_STACK_REGION     =   "Stack";
const std::string	CPHF_S_LABEL_STACK_REGION   =   "STCKR";

const std::string	CPHF_LABEL_ESP              =   "Stack Pointer"; //not explicitly implemented
const std::string	CPHF_S_LABEL_ESP            =   "STCKP";

const std::string	CPHF_LABEL_EPI              =   "Instruction Pointer";
const std::string	CPHF_S_LABEL_EPI            =   "INSTP";

#endif
